targets/psa_switch/psa_switch.cpp

```diff
@@ -313,9 +301,21 @@ PsaSwitch::ingress_thread() {
 
     phv = packet->get_phv();
 
+    /* PSA wants us to resubmit the instance of the packet BEFORE ingress
+       processing, so store the "parsed headers" prior to parsing. Feels a bit
+       hacky though, same as in simple_switch.cpp
```

Comment:
The comment 'so store the "parsed headers" prior to parsing' leads me to wonder if there is some confusion here.

The contents of the "parsed headers" prior to parsing should _always_ be the same: all headers invalid.

I do not know the precise effect of packet->save_buffer_state(), and whether what it is doing is only saving the state of the parsed headers, or perhaps is also doing other things.

Reply:
Except for that one comment of mine, which is only about the wording of a comment and so not necessary to change for code-functionality reasons, this all looks good to me.  I haven't tested it yet, though.

Reply:
I probably have this wrong. Will look into it to correct the comment

Reply:
From an earlier comment by Antonin, who knows this code much better than I: https://github.com/p4lang/behavioral-model/pull/773#discussion_r281438818

it seems that perhaps one of the things stored in the packet object is the number of bytes "consumed" by the parser, i.e. bytes that are extract'ed, or advance'd past.  Restoring that number of bytes back to 0 must be done for resubmitted packets (and also ingress-to-egress cloned packets -- not handled yet in PSA implementation), so that when the parser code is executed again, it starts at the beginning of the packet again.  That might be what the save_buffer_state() is accomplishing here (or at least the part about saving the offset of 0 -- the restoring it back to the saved offset of 0 is done later with a different call, I believe).

Reply:
thanks - comment has been updated!